rollup/sync_service.go

```diff
@@ -1,641 +1,258 @@
-	}
-	s.Eth1Data = eth1Data
 
-	_, client, err := s.dialEth1Node()
-	if err != nil {
-		return fmt.Errorf("Cannot dial eth1 nodes: %w", err)
-	}
-	s.ethrpcclient = client
-	s.ethclient = client
-	s.logClient = client
+		bn := service.GetLatestL1BlockNumber()
+		ts := service.GetLatestL1Timestamp()
+		log.Info("Initialized Latest L1 Info", "blocknumber", bn, "timestamp", ts)
 
-	err = s.verifyNetwork()
-	if err != nil {
-		return fmt.Errorf("Wrong network: %w", err)
-	}
-	// Resolve addresses and set them globally
-	err = s.resolveAddresses()
-	if err != nil {
-		return fmt.Errorf("Error resolving addresses: %w", err)
-	}
-	// Bind to the contracts
-	err = s.bindContracts()
-	if err != nil {
-		return fmt.Errorf("Error binding to contracts: %w", err)
-	}
-	// Check the sync status of the eth1 node
-	err = s.checkSyncStatus()
-	if err != nil {
-		return fmt.Errorf("Bad sync status: %w", err)
+		var i, q string
+		index := service.GetLatestIndex()
+		queueIndex := service.GetLatestEnqueueIndex()
+		if index == nil {
+			i = "<nil>"
+		} else {
+			i = strconv.FormatUint(*index, 10)
+		}
+		if queueIndex == nil {
+			q = "<nil>"
+		} else {
+			q = strconv.FormatUint(*queueIndex, 10)
+		}
+		log.Info("Initialized Eth Context", "index", i, "queue-index", q)
 	}
 
-	// Set the initial values of the `LatestL1BlockNumber`
-	// and `LatestL1Timestamp`
-	err = s.initializeLatestL1()
-	if err != nil {
-		return fmt.Errorf("Cannot set latest L1: %w", err)
+	// The sequencer needs to sync to the tip at start up
+	// By setting the sync status to true, it will prevent RPC calls.
+	// Be sure this is set to false later.
+	if !service.verifier {
+		service.setSyncStatus(true)
 	}
 
-	go s.LogDoneProcessing()
-	// Catch up to the tip of the eth1 chain
-	err = s.processHistoricalLogs()
-	if err != nil {
-		return fmt.Errorf("Cannot process historical logs: %w", err)
-	}
+	return &service, nil
+}
 
-	gasLimit, err := s.mgrCaller.GetMaxTransactionGasLimit(&bind.CallOpts{
-		Context: s.ctx,
-	})
+func (s *SyncService) ensureClient() error {
+	_, err := s.client.GetLatestEthContext()
 	if err != nil {
-		return fmt.Errorf("Cannot fetch gas limit: %w", err)
+		return fmt.Errorf("Cannot connect to data service: %w", err)
 	}
-	s.gasLimit = gasLimit.Uint64()
-	log.Info("Setting max transaction gas limit", "gas limit", s.gasLimit)
+	return nil
+}
 
-	go s.Loop()
-	go s.pollHead()
-	go s.ClearTransactionLoop()
+// Start initializes the service, connecting to Ethereum1 and starting the
+// subservices required for the operation of the SyncService.
+// txs through syncservice go to mempool.locals
+// txs through rpc go to mempool.remote
+func (s *SyncService) Start() error {
+	if !s.enable {
+		return nil
+	}
+	log.Info("Initializing Sync Service", "eth1-chainid", s.eth1ChainId)
 
+	// When a sequencer, be sure to sync to the tip of the ctc before allowing
+	// user transactions.
 	if !s.verifier {
-		go s.sequencerIngestQueue()
+		err := s.syncTransactionsToTip()
```

Comment:
This should also sync the pending queue elements to prevent a race condition on startup